File: smart-hub/src/main/kotlin/io/github/tassiLuca/hub/coroutines/application/ThermostatHubManager.kt
package io.github.tassiLuca.hub.coroutines.application

import io.github.tassiLuca.hub.coroutines.core.SensorHealthChecker
import io.github.tassiLuca.hub.coroutines.core.TemperatureEntry
import io.github.tassiLuca.hub.coroutines.core.Thermostat
import kotlinx.coroutines.flow.Flow
import kotlin.coroutines.CoroutineContext
import kotlin.time.Duration.Companion.seconds

/** Manages the thermostat and the temperature sensors. */
class ThermostatHubManager(coroutineContext: CoroutineContext) {

    /** The thermostat. */
    val thermostat = Thermostat(10.seconds, coroutineContext)

    /** The temperature sensors checker. */
    val temperatureSensorsChecker = SensorHealthChecker<TemperatureEntry>()

    /** Runs the thermostat and the temperature sensors. */
    suspend fun run(sensorSource: Flow<TemperatureEntry>) {
        thermostat.run()
        sensorSource.collect {
            thermostat.react(it)
            temperatureSensorsChecker.react(it)
        }
    }
}

File: smart-hub/src/main/scala/io/github/tassiLuca/hub/application/ThermostatHubManager.scala
package io.github.tassiLuca.hub.application

import gears.async.{Async, AsyncOperations, ReadableChannel}
import io.github.tassiLuca.hub.core.ports.{AlertSystemComponent, DashboardServiceComponent, HeaterComponent}
import io.github.tassiLuca.hub.core.{
  SensorHealthCheckerComponent,
  TemperatureEntry,
  ThermostatComponent,
  ThermostatScheduler,
}
import io.github.tassiLuca.rears.{Controller, bufferWithin}

import concurrent.duration.DurationInt
import scala.language.postfixOps

trait ThermostatHubManager
    extends ThermostatComponent
    with SensorHealthCheckerComponent[TemperatureEntry]
    with HeaterComponent
    with AlertSystemComponent
    with DashboardServiceComponent:
  override val thermostat: Thermostat = Thermostat(ThermostatScheduler.byHour(19))
  override val sensorHealthChecker: SensorHealthChecker = SensorHealthChecker()

  def run(source: ReadableChannel[TemperatureEntry])(using Async, AsyncOperations): Unit =
    thermostat.asRunnable.run
    sensorHealthChecker.asRunnable.run
    Controller.oneToMany(
      publisherChannel = source,
      consumers = Set(thermostat, sensorHealthChecker),
      transformation = r => r.bufferWithin(10.seconds),
    ).run
